CEL library extensions for IP Address and CIDR network parsing

JoelSpeed opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

This PR adds new extensions to the CEL extension library for Kubernetes to allow users to validate common IP address comparisons.

The intention of this PR is to allow users to create CEL rules that will allow them to determine:
* `isIP(<string>)` - Is the string a valid ipv4/ipv6 addrss
* `isCIDR(<string>)` - Is the string a valid ipv4/ipv6 subnet in CIDR notation
* `ip.isCanonical(<string>)` - Returns whether the IP address is in its canonical format
* `ip(<string>).family()` - Returns an integer (4 or 6) based on IP address network family
* `ip(<string>).isUnspecified()` / `ip(<string>).isLoopback()` / `ip(<string>).isLinkLocalMulticast()` / `ip(<string>).isLinkLocalUnicast()` / `ip(<string>).isGlobalUnicast()`- Boolean values exposing `netip.Addr` methods for special IP addresses
* `cidr(<string>).containsCIDR(cidr(<string>))` - is a particular CIDR contained within another CIDR
* `cidr(<string>).containsIP(ip(<string>))` - is a particular IP contained within the CIDR
* `cidr(<string>).containsCIDR(<string>)` - is a particular CIDR contained within another CIDR
* `cidr(<string>).containsIP(<string>)` - is a particular IP contained within the CIDR
* `cidr(<string>).ip()` - Returns the IP representation of the CIDR address, can then be manipulated using `ip()` members
* `cidr(<string>).masked()` - Returns the masked representation of the CIDR object, can be used to compare whether the original CIDR was masked or not
* `cidr(<string>).prefixLength()` - Returns the prefix length of the CIDR network

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add CEL library for IP Addresses and CIDRs. This will not be available for use until 1.31.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

JoelSpeed committed:
Add IP address CEL extension

This adds new CEL functions to the library for validating if a string is an IP address,
and, if it can be parsed as an IP address, adds additional accessors to get properties
of the IP address.

JoelSpeed committed:
Add CIDR network CEL extension

This adds new CEL functions to the library for validating if a string is a CIDR notation.

This will work in conjunction with the IPAddr to allow checking if an IPAddr exists within a particular network.

cici37 commented:
Hi Joel, thanks for raising the PR! Just wanna mention a couple TODOs before going into the details. 
1. Adding the library into the dark mode by specifying the introduced version here: https://github.com/kubernetes/kubernetes/blob/8254569d50d6bbe8d60bcdfb16eac337f6e60514/staging/src/k8s.io/apiserver/pkg/cel/environment/base.go#L49
2. The cost estimation which you might be already aware of. Just wanna mention that we would also prefer to have the cost related tests in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/cel/library/cost_test.go for consistency.
3. The test coverage. We might have to add further tests after the library is general available due to the way we introduce the new library. But unit tests are already appreciated. 
4. The documentation. We would have to add documentation a release after(in 1.31 after the library are general available) hence I would suggest to have a task tracked in https://github.com/kubernetes/website.

Thank you for the quick implementation!

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/library/ip_addrs.go:
```diff
@@ -0,0 +1,151 @@
+//	isIPAddr('127.0.0.256') // returns false
+//	isIPAddr(':::1') // returns false
+//
+// isIPv4 / isIPv6 / isInCIDR
```
remove `isInCIDR`?  Looks like this is handled by the below `containsIPAddr`.

JoelSpeed replied:
I started by writing out the docs for what I thought I wanted in this, and changed my mind half way through putting the cidr to ip check on the network address side, so yep, forgot to remove this

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/library/network_addrs.go:
```diff
@@ -0,0 +1,141 @@
+	apiservercel "k8s.io/apiserver/pkg/cel"
+)
+
+// NetworkAddr provides a CEL function library extension of network address parsing functions.
```
nit: It might just be me.. But I didn't immediately realize this type was for CIDR blocks. Should we somehow highlight in this description?

JoelSpeed replied:
Yep, we can either update the docs or change the name of this. Will see if I can find someone with better networking experience than I have to talk through different naming suggestions and come back to this

jpbetz reviewed (commented):
This is what I was hoping for!

Do other libraries like this typically provide any `CIDR.in(cidr)` or `CIDR.disjoint(cidr)`? Worth adding? (I don't mean this to be a leading question, I'm genuinely curious if this is needed or not)

JoelSpeed commented:
> Do other libraries like this typically provide any CIDR.in(cidr) or CIDR.disjoint(cidr)? Worth adding? (I don't mean this to be a leading question, I'm genuinely curious if this is needed or not)

Good question, will have a look at what else is around. The netip.Prefix (that I'm using to represent the network underneath) has an [Overlaps](https://pkg.go.dev/net/netip#Prefix.Overlaps) method which could be used in the negative to create a disjoint, so maybe this is something we want to expose.

Looking at how that's implemented, I suspect we could also create a `a.containsNetworkAddr(b)`(if we are keeping the NetworkAddr term at least) based on masking both a and b using a's `Bits()` and then checking that once masked they addresses are equal and that the bits of b are greater than the bits of b ([demo](https://go.dev/play/p/KU4-H-LNRzi))

aojea commented:
we had a bit of a drama with IP addresses https://github.com/kubernetes/kubernetes/issues/108074 due to change in the golang parser to mitigate a CVE, so it made validation stricter but we have decided to be compatible with the old representation https://github.com/kubernetes/kubernetes/pull/104368 forking the parsers https://github.com/kubernetes/utils/pull/207

From the existing PR I can see this is using the "new" net/netip parsers , that is definitively where we want to go https://github.com/kubernetes/kubernetes/issues/106161 , (although @danwinship found a  bug there recently https://github.com/golang/go/issues/63850 , but I think that is ok)

The question is that these new extensions will consider invalid IPs with leading zeros, that are valid for the existing objects, @thockin @liggitt WDYT? sooner or later we need to stop using the forked parsers and is unlikely that these ips with leading zeros can still be consumed , at least not from golang software

JoelSpeed commented:
> The question is that these new extensions will consider invalid IPs with leading zeros, that are valid for the existing objects, @thockin @liggitt WDYT? sooner or later we need to stop using the forked parsers and is unlikely that these ips with leading zeros can still be consumed , at least not from golang software

I assume the question here is more about whether we are ok having two different versions of validation rather than introducing something stricter in this new validation? No one _should_ be adding this new CEL to an already shipped API field as that would in theory represent a tightening of validation and be a breaking change, and I don't think we are proposing for anyone to replace any existing validation with this either.

@aojea Are you thinking perhaps that this is something that should wait for ratcheting validation?

danwinship commented on staging/src/k8s.io/apiserver/pkg/cel/library/ip_addrs.go:
```diff
@@ -0,0 +1,151 @@
+		return types.MaybeNoSuchOverloadErr(arg)
+	}
+
+	addr, err := netip.ParseAddr(s)
```
So I've been pondering `netip` wrappers / better k8s IP utilities. https://github.com/danwinship/kubernetes/commits/ipfamily

My current attempt at a `ParseIP` function is [here](https://github.com/danwinship/kubernetes/commit/9bc20ff8cc4ff75d41b4e256889ba1b8c8134828#diff-311b2fd92f5d259a223ae13ba5792a4f9a7e50366815553214285b3924db31faR48-R61). As compared to this:

1. It returns an error if `addr.Zone() != ""`, which we definitely want here. (`netip.ParseAddr` allows strings such as `fe80::1234%eth0` which you need to be able to do sometimes, but which there are probably no good use cases for having in kubernetes API objects). (`net.ParseIP` does not allow zones in IPs, so this is the more-compatible-with-existing-code option anyway.)
2. It returns an error if `addr.Is4In6()`, i.e. `"::ffff:192.168.0.1"`, which again, there are some contexts where it's useful to be able to treat IPv4 addresses as IPv6 addresses, but kubernetes API objects are not such a context, so we should reject those too. (Most existing API objects *do* allow 4-in-6 addresses, and this is terrible, and there's probably at least one CVE lurking in there somewhere.)
3. It [overrides `netip.ParseAddr`'s error messages in some cases](https://github.com/danwinship/kubernetes/commit/9bc20ff8cc4ff75d41b4e256889ba1b8c8134828#diff-311b2fd92f5d259a223ae13ba5792a4f9a7e50366815553214285b3924db31faR114-R131), because `netip.ParseAddr`'s error messages are not especially helpful in the case where a user misunderstood what sort of data they were supposed to pass and passed something else (like an IP:port or a CIDR) instead. Actually, ParseAddr's error messages are pretty non-user-friendly in general and we might want to just return a static "invalid IP address" in any case that isn't one of the other mistakes.

(Assuming that branch or something like it eventually merges, then we can later make the functions here use the more generic functions there. But we should at least get the parsing right now.)

danwinship commented on staging/src/k8s.io/apiserver/pkg/cel/library/ip_addrs.go:
```diff
@@ -0,0 +1,151 @@
+//	ipAddr('127.0.0.1').isIPv4Addr() // returns true
+//	ipAddr('::1').isIPv6Addr() // returns true
+//	ipAddr('127.0.0.1').isIPv6Addr() // returns false
+//	ipAddr('::1').isIPv4Addr() // returns false
```
I would add:

```
isCanonical

Returns true if an address is in canonical form. There is exactly one canonical form
for every IP address, so fields containing IPs in canonical form can just be treated
as strings when checking for equality or uniqueness.

Examples:

     ipAddr('127.0.0.1').isCanonical() // returns true; all valid IPv4 addresses are canonical
     ipAddr('2001:db8::abcd').isCanonical() // returns true
     ipAddr('2001:DB8::ABCD').isCanonical() // returns false
     ipAddr('2001:db8::0:0:0:abcd').isCanonical() // returns false
```

in terms of code, `netip.Addr.String()` always returns canonical form, so you'd just have to add another field to `IPAddr` indicating whether the input string equals `Addr.String()`.

danwinship replied:
oh, and likewise for CIDRs (where a canonical CIDR has a canonical IP address, and a prefix length with no leading 0s)

danwinship replied:
`ServiceCIDR` requires canonical form. :slightly_smiling_face: 

Arguments for requiring canonical form:
- allows code/clients to just treat the fields as strings when checking equality/uniqueness
- helps ensure reconciliation loops don't try to re-reconcile already-correct values

Arguments for not requiring canonical form:
- "I want to directly use the output of some other program that outputs non-canonical IPs without having to fix them up".
- "I DON'T LIKE LOWERCASE LETTERS"

thockin replied:
As a rule, we do not modify user input. If they gave us something in the API. It's either valid or it's not valid but we don't change it. I don't think you're really suggesting that we would reject IPs in non-conical form though, right?

aojea replied:
> ServiceCIDR requires canonical form.

Because is an specific implementation and you can specify the input requirements, but I think that in a generic library like this you can not make this assumption

JoelSpeed replied:
> As a rule, we do not modify user input. If they gave us something in the API. It's either valid or it's not valid but we don't change it. I don't think you're really suggesting that we would reject IPs in non-conical form though, right?

Currently `ip` allows non-canonical and users of the CEL library could make their own choice for their own use case, `isIP(<string>) && ip(<string>).isCanonical().

> Require that net == net.Masked(), to limit it to the subnet/mask case rather than the IP-with-prefix-length case.

But for `cidr` @danwinship suggested that we should reject the non-canonical form, but we could add a `masked()` there so users could make their own choice, `isCIDR(<string>) && cidr(<string>) == cidr(<string>).masked()` (I'll have to work out how to make the `==` operator work for the custom library).

I don't have a strong opinion on the networking side of this but can see the API side argument of leaving the library less-opinionated and allowing user choice, so unless @danwinship you think this would be a security issue, are you happy with the library being slightly less opinionated on cidr being canonical form?

danwinship replied:
> As a rule, we do not modify user input.

Right. Wasn't suggesting that.

> I don't think you're really suggesting that we would reject IPs in non-canonical form though, right?

I'm not saying we should change validation of existing APIs to require canonical form, but I think there's a good argument to be made for requiring canonical form in _new_ APIs. Mostly because (a) it has advantages, (b) it doesn't really have disadvantages.

> I think that in a generic library like this you can not make this assumption

Right, that's why I was suggesting it as a separate function; as Joel said, the CRD author can choose whether to require "valid IP" or "valid IP in canonical form".

> But for `cidr` @danwinship suggested that we should reject the non-canonical form

No, I said we should consider IP-with-prefix-length to be not _valid_ for our "CIDR" type. "Canonical" is still a separate consideration. ("`2001:db8::1/64`" is not valid, "`2001:DB8::/64` is valid but not canonical, "`2001:db8::/64`" is valid and canonical).

danwinship commented on staging/src/k8s.io/apiserver/pkg/cel/library/ip_addrs.go:
```diff
@@ -0,0 +1,151 @@
+		return types.MaybeNoSuchOverloadErr(arg)
+	}
+
+	_, err := netip.ParseAddr(s)
```
(needs the same checks as `stringToIpAddr`, whatever gets added there)

liggitt commented:
it's unclear to my why we would add CEL functions for these over the ipv4 / ipv6 openapi format specifiers

danwinship commented on staging/src/k8s.io/apiserver/pkg/cel/library/network_addrs.go:
```diff
@@ -0,0 +1,141 @@
+	apiservercel "k8s.io/apiserver/pkg/cel"
+)
+
+// NetworkAddr provides a CEL function library extension of network address parsing functions.
```
I would just call this "CIDR". It matches `net.ParseCIDR` and I'm pretty sure it's the term we most commonly use. Admittedly, it's not super accurate: "CIDR" ("Classless Inter-Domain Routing") technically refers to the entire concept of breaking up IPv4 subnets on arbitrary boundaries, and the syntax is more properly "CIDR notation", except also, CIDR notation can be used either to refer to subnets/masks ("`1.2.3.0/24`") or individual IPs ("`1.2.3.4/24`" meaning "the IP `1.2.3.4` and by the way it's on a `/24` subnet"), but we only want this to mean the former. But I don't think there's really any standard simple name for that...

(However, "network address" feels definitely bad to me, because `netlink.Addr` is used in the CNI code to mean the "`1.2.3.4/24`" case.)

danwinship commented on staging/src/k8s.io/apiserver/pkg/cel/library/network_addrs.go:
```diff
@@ -0,0 +1,141 @@
+//	networkAddr('192.168.0.0/16') // returns an IPv4 address with a CIDR mask
+//	networkAddr('::1/128') // returns an IPv6 address with a CIDR mask
+//	networkAddr('192.168.0.0/33') // error
+//	networkAddr('::1/129') // error
```
I'd include

```
     networkAddr('192.168.0.1/16') // error, because there are non-0 bits after the prefix
```

There are no APIs in `k8s.io` that _want_ an IP-with-prefix-length value rather than a subnet/mask value, but there are currently APIs that incorrectly allow them (https://github.com/kubernetes/kubernetes/pull/94484). We should forbid them here. If someone later needs an IP-with-prefix-length value (Multi-Network maybe?) they can add new functions for that.

JoelSpeed replied:
That is to say that, when we parse a CIDR in this string using `netip.ParsePrefix`, we want to also make sure that the result `Addr()` is the same as `Masked().Addr()` right? Masking the address we got it should not change the address

danwinship replied:
right

danwinship commented on staging/src/k8s.io/apiserver/pkg/cel/library/network_addrs.go:
```diff
@@ -0,0 +1,141 @@
+//
+// containsIPAddr
+//
+// - containsIPAddr: Returns true if a network address contains an IP address. Should be passed an IP address parsed with ipAddr.
```
just "containsIP" ?

FWIW, (a) we might want `containsCIDR` too, and (b) it's possible that inverting them (having `IPAddr.isInCIDR` like you'd thought originally, and `NetworkAddr.isInCIDR`) might be more useful (or maybe have it available both ways?). The one place in core k8s I can think of right now where we might use this is in [NetworkPolicy `ipBlock` validation](https://github.com/kubernetes/kubernetes/blob/v1.28.4/pkg/apis/networking/validation/validation.go#L239), where given:

```
- ipBlock
    cidr: 192.168.1.0/24
    except:
      - 192.168.1.1/32
```

you want to check that each `except` value `isInCIDR(cidr)`, which seems easier to check in that direction than in the reverse?

JoelSpeed replied:
> just "containsIP" ?

I like `containsIPAddr` because it's then obvious that you should be using the `ipAddr` function to parse the IP before you pass it into the function. If we decide to change things around this may go away though.

---

I think it doesn't really matter, for both of these, you need to parse the object and validate it. So you'd either have (I'm renaming to CIDR so examples using new names)
```
isCIDR(self.cidr) && self.except.all(x, isIPAddr(x) && cidr(self.cidr).containsIPAddr(x))
```
or
```
isCIDR(self.cidr) && self.except.all(x, isIPAddr(x) && ipAddr(x).isInCIDR(cidr(self.cidr)))
```
Unless there's an easier way to write one of those?

The benefit I guess of doing `ipAddr.isInCIDR` would be then you have `isInCIDR` on both, rather than having `containsCIDR` and `containsIPAddr` both on the `cidr` object.

Will look to other examples to see if there's a preferential style 

--- 

I think we want to have only 1 way to represent an expression, so having both I'm a downvote on, but which way we do this (`cidr.containsIPAddr`/`cidr.containsCIDR`) vs (`ipAddr.isInCIDR`/`cidr.isInCIDR`) I'm not strongly opinionated on

danwinship replied:
ah, I was thinking only of simpler cel, where the rule would be on the `except` field rather than being able to express it at a higher level

danwinship commented on staging/src/k8s.io/apiserver/pkg/cel/library/network_addrs.go:
```diff
@@ -0,0 +1,141 @@
+// Examples:
+//
+// networkAddr('192.168.0.0/24').containsIPAddr(ipAddr('192.168.0.1')) // returns true
+// networkAddr('192.168.0.0/24').containsIPAddr(ipAddr('192.168.1.1')) // returns false
```
This syntax seems like it will _always_ require you to call `ipAddr()` on the argument to `containsIPAddr` wouldn't it? It would be simpler to just have `containsIPAddr` take a string

JoelSpeed replied:
Yeah it does, this was inspired by other CEL extension like the quantity CEL that has a similar pattern. I'll let @cici37 chime in here, I think this is the expected pattern, but if taking a string is simpler then we can do that too

danwinship commented on staging/src/k8s.io/apiserver/pkg/cel/network_addr.go:
```diff
@@ -0,0 +1,83 @@
+
+// NetworkAddr provides a CEL representation of an network address.
+type NetworkAddr struct {
+	*netip.Prefix
```
again, don't need the `*`

danwinship commented on staging/src/k8s.io/apiserver/pkg/cel/network_addr.go:
```diff
@@ -0,0 +1,83 @@
+)
+
+// ConvertToNative implements ref.Val.ConvertToNative.
+func (d NetworkAddr) ConvertToNative(typeDesc reflect.Type) (interface{}, error) {
```
Again, possibly support conversion to `*net.IPNet`, though this time [it's a bit more complicated](https://github.com/danwinship/kubernetes/commit/51ce48349fb5fd76ea5e8b4a05b8f6848796daab#diff-8ccdf299272f29f007cdba9f7275f3582de759dfb6d1e3f967ce592c3c67a4dbR84-R108). (Though you can remove some of that error checking if you already validated the `netip.Prefix`.)

danwinship commented on staging/src/k8s.io/apiserver/pkg/cel/network_addr.go:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	equalAddr := d.Prefix.Addr().Compare(otherD.Prefix.Addr()) == 0
+	equalBits := d.Prefix.Bits() == otherD.Prefix.Bits()
```
again, can compare with `==`

danwinship commented on staging/src/k8s.io/apiserver/pkg/cel/library/network_addrs.go:
```diff
@@ -0,0 +1,141 @@
+		return types.MaybeNoSuchOverloadErr(arg)
+	}
+
+	net, err := netip.ParsePrefix(s)
```
[My version](https://github.com/danwinship/kubernetes/commit/9bc20ff8cc4ff75d41b4e256889ba1b8c8134828#diff-f4b3c1b088debfb8229ad7432d5f86310c531d3c9c4a2e8f3cdba9bf665f01c0R50-R73). In particular:

1. Definitely don't allow 4-in-6 (e.g. "`::ffff:1.2.3.0/24`"). (`net.ParseCIDR` and `netip.ParsePrefix` have incompatible ideas on how to parse 4-in-6 CIDRs anyway.)
2. Require that `net == net.Masked()`, to limit it to the subnet/mask case rather than the IP-with-prefix-length case.

(The code there also has another check to rule out the current golang stdlib wackiness Antonio mentioned, which will be fixed in golang 1.22, plus another error-message-wrapper.)

danwinship reviewed (commented):
Woo! This is great.

aojea commented:
> No one _should_ be adding this new CEL to an already shipped API field as that would in theory represent a tightening of validation and be a breaking change, and I don't think we are proposing for anyone to replace any existing validation with this either.

that answers my question @JoelSpeed , thanks

JoelSpeed commented on staging/src/k8s.io/apiserver/pkg/cel/ip_addr.go:
```diff
@@ -0,0 +1,80 @@
+)
+
+// ConvertToNative implements ref.Val.ConvertToNative.
+func (d IPAddr) ConvertToNative(typeDesc reflect.Type) (interface{}, error) {
```
I'm not sure why there would be conversion to a `net.IP` since this library is intended to be used only within the CEL validation, meaning, IIUC, built in types/anything that already exists with `net.IP` compatibility wouldn't be anywhere near this code

danwinship replied:
ah, sorry, I haven't done anything with CEL extensions and was just kind of guessing at what some of this is for. :slightly_smiling_face: I wasn't sure if this function affected what types you could use for the field when generating structs from a CRD...

JoelSpeed replied:
Nah, nothing to do with the generation, this can be super simple and isolated I think.

JoelSpeed reviewed (commented):
Thanks for the feedback on this one @danwinship, I'm working on those changes and will probably push some updates tomorrow. Where I have responses I've responded inline.

I think [this](https://github.com/kubernetes/kubernetes/pull/121912/files#r1395717510) thread is particularly worth highlighting in case others have thoughts and want to add anything to the vote for one way or the other.

thockin commented:
PR comment says `isNetworkAddr` but code saysd `isCIDR` ?

JoelSpeed commented:
> PR comment says `isNetworkAddr` but code saysd `isCIDR` ?

Got a couple of notes of feedback that NetworkAddr wasn't a great name so have switched to CIDR in the latest commit, will update PR message

thockin reviewed (commented):
I'm assuming the goal here is to make progress towards CEL validation for things like Services and Endpoints and Ingress and Gateway.  As such my comments are mostly around things we do in there.

I see ipAddr.isIPv[46].  Consider a method on both types, such as `family()` or `ipFamily()` which returns whether it is IPv4 or IPv6.  Why?  We have some validations that check that two list elements are of different families (`ipAddr(ips[0]).family() != ip[Addr(ips[1]).family()`) for dual-stack support, but don't care which order.

Consider methods on ipAddr or maybe both that check for "specialness" (e.g. loopback, link-local, multicast).  Why?  We have validations that ensure that things like Services can't be tricked into accessing things the user can't reach themselves.

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/library/ip_addrs.go:
```diff
@@ -0,0 +1,199 @@
+//
+// Examples:
+//
+//	ipAddr('127.0.0.1') // returns an IPv4 address
```
I also think we should not allow it ... if this new validator is added to an existing type, I'd lean on ratcheting validation to deal with letting existing data be updated

danwinship replied:
The consensus a year and a half ago in #108074 was that we should stop accepting invalid (leading 0) values for all newly-created objects / newly-updated fields of existing objects, even where that is technically an API break, because security. I'm pretty sure the only reason we never actually implemented it is because the validation code as currently written makes it hard to do that without also forcing stricter revalidation of *not*-updated fields of existing objects and breaking the world.

aojea replied:
Agree with Dan and Jordan, let's not carry forward the sloppy parsers

JoelSpeed replied:
Since I'm using `netip` to parse the addresses in both the ip and cidr implementations, the library doesn't allow leading zeros, so I just need to update the documentation to reflect that.
An example error from the library:
```
ParseAddr("10.0.0.010"): IPv4 field has octet with leading zero
```

JoelSpeed commented:
> I'm assuming the goal here is to make progress towards CEL validation for things like Services and Endpoints and Ingress and Gateway. As such my comments are mostly around things we do in there.

Yeah that's right. I've personally been dealing with a number of validation issues in both the Ingress space and in Cluster API style use cases where this CEL would be helpful. I know others on this thread have experience in other areas such as CNI and Gateway API so I think there are a few different use cases coming together.

> I see ipAddr.isIPv[46]. Consider a method on both types, such as family() or ipFamily() which returns whether it is IPv4 or IPv6. Why? We have some validations that check that two list elements are of different families (ipAddr(ips[0]).family() != ip[Addr(ips[1]).family()) for dual-stack support, but don't care which order.

I expect we want to keep to "only one way to express a statement" and I think `ipAddr(ips[0]).isIPv4() != ipAddr(ips[1]).isIPv4()` would be logically equivalent. So if we prefer to have a `family()` return a value of some type, we may want to drop these current methods. I'll have to see what other CEL libraries look like to see if there's a preferred style.

> Consider methods on ipAddr or maybe both that check for "specialness" (e.g. loopback, link-local, multicast). Why? We have validations that ensure that things like Services can't be tricked into accessing things the user can't reach themselves.

Good feedback, I saw methods similar to this on the `netip.IP` so they should be relatively straightforward to implement if that's something we think is going to be useful, can add.

thockin commented:
> I expect we want to keep to "only one way to express a statement" and I think `ipAddr(ips[0]).isIPv4() != ipAddr(ips[1]).isIPv4()` would be logically equivalent.

I can deal with that, though `ip.family() == 4` and `ip.isIPv4()` are isomorphic.  I can live with whichever is more consistent with the rest of CEL libs.

> they should be relatively straightforward to implement if that's something we think is going to be useful, can add

Maybe look at how we validate endpoint IPs today and start with that subset?

danwinship commented:
> Consider methods on ipAddr or maybe both that check for "specialness" (e.g. loopback, link-local, multicast). Why? We have validations that ensure that things like Services can't be tricked into accessing things the user can't reach themselves.

Yeah, [ValidateNonSpecial](https://github.com/kubernetes/kubernetes/blob/v1.28.4/pkg/apis/core/validation/validation.go#L6732-L6760)



> > We have some validations that check that two list elements are of different families (ipAddr(ips[0]).family() != ip[Addr(ips[1]).family()) for dual-stack support, but don't care which order.
> 
> I expect we want to keep to "only one way to express a statement" and I think `ipAddr(ips[0]).isIPv4() != ipAddr(ips[1]).isIPv4()` would be logically equivalent.

Historically people have written stuff like `netutils.IsIPv6(foo) == netutils.IsIPv6(bar)` in our go code too. But then other people have a tendency to misread that as saying "assert that foo and bar are both IPv6" rather than "assert that foo and bar are the same IP family". (I had to correct someone getting confused by that in more than one PR review.) So I added [`netutils.IPFamilyOf`](https://github.com/kubernetes/utils/pull/271).

(`netutils.IPFamilyOf` turned out to be not as useful as I hoped though because a lot of the time you want to compare against a `v1.IPFamily`, but `k8s.io/utils` can't import `k8s.io/api`, and you can't even cast between the two sets of values because they have different representations. That's another thing I'm working on with my new IP utils I linked to before.)

Anyway, there are situations where `.isIPv4`/`.isIPv6` are more idiomatic and useful, and situations where `.family` is more idiomatic and useful.

JoelSpeed commented on staging/src/k8s.io/apiserver/pkg/cel/library/cidr.go:
```diff
@@ -0,0 +1,192 @@
+//
+// containsIPAddr / containsCIDR
+//
+// - containsIPAddr: Returns true if a network address contains an IP address. Should be passed an IP address parsed with ipAddr.
```
I had gone with `containsIPAddr` because the method's argument takes an `ipAddr()`, but will check what other libraries are doing and shorten if that seems to be convention. Maybe just to `containsIP` though since we also have `containsCIDR` 🤔 This may be another argument for switching to `isInCIDR` on both the `ipAddr` and `cidr` types, would be consistent

JoelSpeed commented on staging/src/k8s.io/apiserver/pkg/cel/library/ip_addrs.go:
```diff
@@ -0,0 +1,199 @@
+//	isIPAddr('127.0.0.256') // returns false
+//	isIPAddr(':::1') // returns false
+//
+// isIPv4 / isIPv6 / isCanonical
```
I like that suggestion, that sounds like a useful addition. Will think on the name base, I would have though `cidr().ip()` would be more obvious that it converts the cidr to an ip representation and that the `ip()` methods are available on that object

JoelSpeed commented:
> Anyway, there are situations where .isIPv4/.isIPv6 are more idiomatic and useful, and situations where .family is more idiomatic and useful.

@danwinship Is that you arguing for supporting both? These libraries are relatively small so I'm happy to add both but conscious that we try to have 1 way to express a thing, and this would mean there are two ways. If you had to chose only one, which would you choose? `ip(<string>).family() == 4` vs `ip(<string>).isIPv4()` and `ip(<string1>).family() != ip(<string2>).family()` vs `ip(<string1>).isIPv4() != ip(<string2>).isIPv4()`

Looking at the various other libraries that I can see, it seems there's a tendency to provide actual data back (strings or ints) vs bools, the data is more useful I guess since you can then make your own comparison with the built in operators. So I think if we only want 1 of these, I would go for `family()`

JoelSpeed committed:
Add special IP validations to IP CEL type

JoelSpeed commented:
Made a number of changes to this today based on feedback from folks, I think I've covered everything so far use case wise:
* Switch `IPAddr` to just `IP` for sake of brevity, CEL will now be `ip` and `cidr`
* Add `isInCIDR` to both `ip` and `cidr` so that they are consistently named and we don't have `containsIP` and `containsCIDR` on `cidr`
* Added special IP address validations as found used in [ValidateNonSpecialIP](https://github.com/kubernetes/kubernetes/blob/v1.28.4/pkg/apis/core/validation/validation.go#L6732-L6760)
* Switched to `family` on `ip` to return `4` or `6`, users can make their own comparison for IP families based on this
* Removed the requirement for a CIDR to be masked and added `masked` to `cidr` to allow users to `cidr() == cidr().masked()` for their own use case

If we are all in agreement with the names and changes so far I'll squash down what I have (the fixup commits) into something more atomic and move onto the testing and cost estimation part of this

JoelSpeed committed:
Add IP and CIDR libraries to CEL environment for 1.30

thockin commented on staging/src/k8s.io/apiserver/pkg/cel/library/cidr.go:
```diff
@@ -0,0 +1,218 @@
+//
+// - ip: Returns the IP address representation of the CIDR.
+//
+// - masked: Returns the CIDR representation of the network address with a masked prefix.
```
You may need a way to query how many prefix bits and/or host bits - ISTR a few places where we enforce maximum size, e.g.

`cidr.family() == 4 ? cidr.hostbits <= 16 : cidr.hostbits <= 64`

I'm not sure what the CIDR spec names those. :)

@aojea

danwinship replied:
The number after the slash is the "prefix length". I'm not sure there's a standard name for the total length minus prefix length. "Host bits" only makes sense in the context of node cidrs, where it means "the number of bits reserved for this host to use for pod IPs" (but if you run `kcm --cluster-cidr 10.128.0.0/20`, it wouldn't really make sense to say that the cluster cidr has "12 host bits", because 4 of those bits are used in one way and 8 in another way).

FWIW I'm not sure we actually care about validating that anywhere except for service network CIDRs, but if we're going to have a selector to get the IP-address part of a CIDR, then it makes sense to have a selector to get the prefix-length part too..

JoelSpeed replied:
So should we expose this under `prefixLength()` then? Would that be the most appropriate/expected name?

danwinship replied:
sure, `prefixLength` sounds right

thockin commented on staging/src/k8s.io/apiserver/pkg/cel/library/cidr.go:
```diff
@@ -0,0 +1,218 @@
+//
+// - containsCIDR: Returns true if a network address contains a CIDR. Should be passed a CIDR parsed with cidr.
+//
+// - ip: Returns the IP address representation of the CIDR.
```
You may need a way to query how many prefix bits and/or host bits - ISTR a few places where we enforce maximum size, e.g.

`cidr.family() == 4 ? cidr.hostbits <= 16 : cidr.hostbits <= 64`

I'm not sure what the CIDR spec names those. :)

@aojea

JoelSpeed committed:
Add costing estimations for IP and CIDR

JoelSpeed commented:
Ok, latest push includes my first attempt at some cost estimation stuff. Took me a while to work out but I think I got there.

For the functions that parse the IP addresses I took inspiration from the other parsing cost estimates, things like the `url`, basing it length of input string.

The fun one was `is_in_cidr`, in which I concluded that the comparisons are all being done based on the prefix length, and we know that the prefix length will be at most 4 bytes for an ipv4 and 16 bytes for an ipv6, so that lead to some bounding on those comparisons to hopefully achieve some reasonable numbers.

I think with the current design you don't actually need a length limit on the `string` type that you'd put this CEL on, since the ip parser will bail out if you go over the correct length anyway, which I would be 18 chars for ipv4 and 43 chars for an ipv6.

thockin commented on staging/src/k8s.io/apiserver/pkg/cel/library/ip.go:
```diff
@@ -0,0 +1,307 @@
+//	ip('::1').family() == 6 // returns true
+//	ip('127.0.0.1').isCanonical() // returns true; all valid IPv4 addresses are canonical
+//	ip('2001:db8::abcd').isCanonical() // returns true
+//	ip('2001:DB8::ABCD').isCanonical() // returns false
```
This feels weird to me.  I'd expect the conversion from non-canonical `string` into `ip` to accept the non-canonical input but to convert to some optimal form (e.g. `[2]uint64` in go), meaning it would not retain things like this.  IOW: strings representations of IPs can be canonical or not, but the IP itself is "pure".


Are we saying instead that the internal representation is going to retain the string?

I would have expected this to look more like a free-function `isCanonicalIP(string) bool`

Am I thinking of it wrong?

danwinship replied:
ah, yeah, it would have to be more like `wasCanonicalIP` really...

I had suggested it as a function on the IP rather than as a freestanding string function because that matched the other functions, which I guess matches CEL norms.

thockin replied:
So we would track a `bool` indicating "we accepted this IP but it was non-canonical" ?  OK I guess.  I'll leave that to @JoelSpeed to decide the right answer

JoelSpeed replied:
The `isCanonical` check as I've implemented is a member of the `IP` type, but it takes the raw/original input string and compares it to the output of the `netip.Addr.String()` which is always canonical. If they are different, it wasn't canonical.

So I think making it freestanding would mean it still has to parse, doing the same work as `ip()` already, so I'm leaning towards keeping it as a member to minimise the top level exposed syntax in the library extension, but also to change the name to `wasCanonicalIP` as suggested by Dan.

danwinship replied:
> but also to change the name to `wasCanonicalIP` as suggested by Dan.

I'm not sure I was actually suggesting that... it feels weird. You are really asking a question about the string ("is this string the canonical representation of some IP address?"), not a question about the IP address that the string represents. So `isCanonical('2001:db8::abcd')` seems like it makes more sense than `ip('2001:db8::abcd').wasCanonical()`.

JoelSpeed replied:
Ok, I'm happy to make it top level and not a member of the `ip()` but I'd like an ack from someone like @cici37 or @jpbetz who has more experience in this than I do

jpbetz replied:
Top level functions are safe to add.  CEL [is able to distinguish between](https://playcel.undistro.io/?content=H4sIAAAAAAAAA3WRTUsDQQyG%2F0ocD22hticvvRUpiFQtihRhL%2Bluujswk1nmo7WI%2F93MdrVW6GlCknneN8mnqjCimqlruCdPoAPERh5uU4RckhDe549LcB4eXp%2BfYOu8xTgpuOAqeYza8QwKddsUqmA1VvTRegpB0kKdTmFNpnSWILqOfLdYwsrgofYucXVVsLSc57IHZNGN5LGMekfC2MxDILsxB2jdXoxWQLzT3rEljpktssbJACgAicnrrrLXsTnqOmsdw%2BLXHSyR64Q1wVD0R5PspHNzA2uvI8HBJd9ZO00EjUj3TW%2BBOjB6QpBSjoOuKC%2Foz%2F7GFxbYU1aZDIOXxINuih2ahBfEsUbNIf47UA9a9PPnf6Uzhsp8GXBbIaBtDYXeWGj18Wqn%2Bw1%2FgpH6%2BgakwBpPEQIAAA%3D%3D) function names and variable identifiers.  If you add one, consider prefixing the function name with a scope like how the recently added [`sets.contains`](https://github.com/google/cel-go/blob/master/ext/sets.go) does.

thockin commented on staging/src/k8s.io/apiserver/pkg/cel/library/ip.go:
```diff
@@ -0,0 +1,307 @@
+//	isIP('127.0.0.256') // returns false
+//	isIP(':::1') // returns false
+//
+// family / isCanonical / isInCIDR / isUnspecified / isLoopback / isLinkLocalMulticast / isLinkLocalUnicast
```
I don't understand why this comment line comes befoe the "Returns true if..."

JoelSpeed replied:
I was following the pattern on the URL library which has something similar, except in this case, that line hasn't been updated. I think we can probably drop it and rely just on the individual descriptions of each function I've added below

thockin commented on staging/src/k8s.io/apiserver/pkg/cel/library/ip.go:
```diff
@@ -0,0 +1,307 @@
+// - isUnspecified: returns true if the IP address is the unspecified address.
+// Either the IPv4 address "0.0.0.0" or the IPv6 address "::".
+//
+// - isLoopback: returns true if the IP address is the loopback address.
```
s/the loopback/a loopback

jpbetz commented:
> I think with the current design you don't actually need a length limit on the `string` type that you'd put this CEL on, since the ip parser will bail out if you go over the correct length anyway, which I would be 18 chars for ipv4 and 43 chars for an ipv6.

Great, yes, this is the direction I was hoping we could take to define cost estimation.

thockin reviewed (commented):
Sorry to nit-pick :)

JoelSpeed commented on staging/src/k8s.io/apiserver/pkg/cel/library/ip.go:
```diff
@@ -0,0 +1,307 @@
+// one canonical form for every IP address, so fields containing IPs in canonical
+// form can just be treated as strings when checking for equality or uniqueness.
+//
+// - isInCIDR: returns true if the IP address is contained in the CIDR.
```
No worries about the nit picking, I understand that once set these can't be changed so we need to get this right!

I've been stuck in Go for so long, that I forgot overloads are a thing. I didn't like the `contains` idea originally because I was thinking we would need two contains methods , one for IP, one for CIDR. But it turns out CEL supports overloading of methods so, we can just create one `contains` with two implementations, and have the user facing side accept either an IP or a CIDR.

Would you be happy with that UX?

thockin replied:
Is overloading "normal" in CEL?  I think we should follow as many normal conventions as possible :)

I am fine with cidr.contains() or cidr.containsIP() + cidr.containsCIDR or (last pref :) ip.isInCIDR().  You probably know CEL norms better than me.

JoelSpeed replied:
I can see overloads in the authz, quantity and regex libraries. Each has at least one method that has the option to take different arguments of different types. So it looks like there is precedence.

Given that, we could make this work with three overloads, `contains(<string>)`, `contains(ip(<string>))` and `contains(cidr(<string>))` where the raw string case would have to check if the string is an IP or CIDR first before delegating to the appropriate contains.

I'd probably want @cici37 or @jpbetz to sign off on the overloading like this

jpbetz replied:
CEL overloading is quite comprehensive and heavily relied on.

I'm supportive of using overloads where the function's declared purpose is the same across all overloads.  I would try to avoid overloads where the documentation will have to define distinct semantics for each possible overload.

In this case, I think I'm in favor of a `containsIP()` function overloaded to take either a string or an IP, and a separate `containsCIDR()` function overloaded to take either a string or a CIDR.

jpbetz replied:
cc @TristonianJones for any additional thoughts

TristonianJones replied:
Meeting the user where they are is great for helping users be successful in writing an expression that 'just works'; However, this can lead to suboptimal performance when the overloads handle a lot of things which would be better to do once, e.g. parse the ip and then perform operations on it.

If you force the user to convert the string to an ip, then you'll get code which is more exact, but not necessarily more clear in its intent.

Since you have cost tracking to encourage optimization, then it seems reasonable to weight the convenience overloads as more costly, but keep them to increase the odds that the expression 'just works'

danwinship commented on staging/src/k8s.io/apiserver/pkg/cel/library/ip.go:
```diff
@@ -0,0 +1,307 @@
+// - isLoopback: returns true if the IP address is the loopback address.
+// Either an IPv4 address with a value of 127.x.x.x or an IPv6 address with a value of ::1.
+//
+// - isLinkLocalMulticast: returns true if the IP address is a link-local multicast address.
```
> "special" is not a standard defintion, so it's not a great candidate for this CEL library (I assume). So where SHOULD it go? "Special" is a kube-centric, definition, though.

This PR is kube-centric. (It's not changing "core CEL" or whatever, it's defining kube-specific extensions.)

But also, `ValidateNonSpecialIP` is nearly identical to `net.IP.IsGlobalUnicast` / `netip.Addr.IsGlobalUnicast`. (`ValidateNonSpecialIP` allows some broadcast and multicast IPs that `IsGlobalUnicast` doesn't, but that's probably a bug.) So we could just call it `isGlobalUnicast`, using the same checks as the golang stdlib, and that would make it less ambiguous what it's supposed to mean.

Currently we only use `ValidateNonSpecialIP` for service externalIPs, endpoint IPs, and endpointslice IPs. It probably ought to be used more than that...

thockin replied:
IsGlobalUnicast seems OK, but we need to make sure it's not a breaking change :)

JoelSpeed replied:
> Or (just thinking ouot loud) is there a way to define "local" functions that get loaded for some evaluations? e.g. in my types.go, somewhere near the top:

To my knowledge, this doesn't exist (yet?). I've not seen any talk of it in the KEP either.

I can add an `IsGlobablUnicast`. I don't think it would be breaking since nothing is using this now, all of this validation is net new. The breaking change would be changing something that uses `ValidateNonSpecialIP` to this without accounting for the differences between them. Are we expecting anyone to do that or is the helper only used on built in types?

danwinship replied:
Tim is worrying about the bright shiny future where we drop all the golang validation and replace it all with CEL. If that happens, we will probably need to add an `isLegacyNonSpecialIP()` method specifically for the objects that are currently using `ValidateNonSpecialIP`, because there's no other way to express *exactly* what that function currently checks unless we export every single existing IP-testing function (`IsLinkLocalMulticast` etc).

JoelSpeed commented on staging/src/k8s.io/apiserver/pkg/cel/library/cidr.go:
```diff
@@ -0,0 +1,218 @@
+//
+//	cidr('192.168.0.0/16') // returns an IPv4 address with a CIDR mask
+//	cidr('::1/128') // returns an IPv6 address with a CIDR mask
+//	cidr('192.168.0.0/33') // error
```
> What does an error mean for the expression evaluator? Is it coverted to "false" or something else?

Any error in parsing is returned to the user as an error and fails the validation, rejecting the request. This means you wouldn't be able to conditionally have logic such as `isCIDR(self) ? cidr(self).bits() == 32 : isIP(self)` to check that if a CIDR is passed then it would be a single address. Not sure you'd want to use that, but this would error if I replaced `isCIDR` with just `cidr`.

> I ask because isCIDR(s) seems OK, but maybe not needed if cidr(s) would produce an equivalent result. It seems like every time we would check isCIDR we also need the result AS a CIDR, so we would parse once to satisfy is and again to convert.

I have followed the patterns set out by the other libraries, such as the URL library. In all the examples I've seen, the parse happens many times. For example, the [CEL playground](https://playcel.undistro.io/?content=H4sIAAAAAAAAA31RW0vDMBT%2BK4cOdgHX6JsUhoIMBPcwFd%2FycpqettE0iUk6Nsf%2Bu%2BnWTfewQR6Sj%2B%2BWc7ZJgQGTLNlyDcATk3%2BSCDzJYA90UO2o7IB4C8H6jLHWk8ssev9Ia2ysolSYJru%2FZRZD%2FfDdktvMVqgGpcOqIR3tOq8d1%2FEkNwmtrSPvpdExlzGYH0w8mBJe2pycphBfH28LeJovQMncodtAqDGA9IArlApzRSB1BAmswk3lTKuLlOvod6z5dfJKpWGFEZ7Fr5AjLSj%2BQepqilYyQYoN%2FrjT1qlpn8k119LHIuPDXNJuFhPgejiESDtD04rCu6ipofEEZjMY7WuMLpOfjQ899XyMo6sSjaeEf7LLmqVxx5hr1nMv0FKxjBvs2fttXha8dmseT1Ivf%2Fo%2Bd3G7u1%2Bd0weKUQIAAA%3D%3D) example parses the URL 7 times.

As far as I understand there is no way to declare a variable or or optimise this so that we only parse once for complex expressions involving these.

JoelSpeed commented:
I've squashed some of the earlier commits down since I don't think we need that history anymore, and I've addressed a bunch of feedback, but did miss [one](https://github.com/kubernetes/kubernetes/pull/121912#discussion_r1397679752) so I still owe that one.

But, summary of changes from today are adding `prefixLength` to `CIDR`, moving all the `isInCIDR` logic to `contains`, adding `isGlobalUnicast`.

thockin commented on staging/src/k8s.io/apiserver/pkg/cel/library/cidr.go:
```diff
@@ -0,0 +1,274 @@
+// - contains: Returns true if a the CIDR contains the given IP address or CIDR.
+// The CIDR must be an IPv4 or IPv6 subnet address with a mask.
+// The IP address must be an IPv4 or IPv6 address.
+// May take either a string, IP address, or CIDR as an argument.
```
is string really needed?  Do we have places where we have a string and don't know if it is a CIDR or an IP?  @danwinship @aojea

aojea replied:
I don't understand it either , string as any random array of characters?

JoelSpeed replied:
Looking at other threads, the most recent suggestion is to move back to `containsIP` and `containsCIDR`, at which point, having the string is helpful but not required, so will keep the string as I make that move.

thockin commented on staging/src/k8s.io/apiserver/pkg/cel/library/cidr.go:
```diff
@@ -0,0 +1,274 @@
+// cidr('192.168.0.1/24').masked() // returns cidr('192.168.0.0/24')
+// cidr('192.168.0.0/24') == cidr('192.168.0.0/24').masked() // returns true, CIDR was already in canonical format
+// cidr('192.168.0.1/24') == cidr('192.168.0.1/24').masked() // returns false, CIDR was not in canonical format
+// cidr('192.168.0.0/16').preefixLength() // returns 16
```
typo: preefix

thockin commented on staging/src/k8s.io/apiserver/pkg/cel/library/ip.go:
```diff
@@ -0,0 +1,302 @@
+//
+// - family: returns the IP addresses' family (IPv4 or IPv6) as an integer, either '4' or '6'.
+//
+// - wasCanonical: returns true when the IP Address was in its canonical format. There is exactly
```
"...when the string used to create the IP Address was..." ?

thockin reviewed (commented):
Other than a few outstanding nits, this LGTM.  No promises we won't need more later :)

TristonianJones commented on staging/src/k8s.io/apiserver/pkg/cel/ip.go:
```diff
@@ -0,0 +1,92 @@
+}
+
+// Value implements ref.Val.Value.
+func (d IP) Value() interface{} {
```
nit: prefer `any` to `interface{}`

TristonianJones commented on staging/src/k8s.io/apiserver/pkg/cel/ip.go:
```diff
@@ -0,0 +1,92 @@
+}
+
+var (
+	IPObject        = decls.NewObjectType("kubernetes.IP")
```
nit: IPObject is unused.

TristonianJones commented on staging/src/k8s.io/apiserver/pkg/cel/ip.go:
```diff
@@ -0,0 +1,92 @@
+var (
+	IPObject        = decls.NewObjectType("kubernetes.IP")
+	ipAddrTypeValue = types.NewTypeValue("kubernetes.IP")
+	IPType          = cel.ObjectType("kubernetes.IP")
```
If the `ip` type doesn't have any accessible fields, prefer to treat it as an opaque type whose properties can only be accessed via accessors, e.g. `cel.OpaqueType("kubernetes.IP")`

TristonianJones commented on staging/src/k8s.io/apiserver/pkg/cel/ip.go:
```diff
@@ -0,0 +1,92 @@
+
+var (
+	IPObject        = decls.NewObjectType("kubernetes.IP")
+	ipAddrTypeValue = types.NewTypeValue("kubernetes.IP")
```
Prefer using `IPType` since it's a type value that implements `ref.Type`

TristonianJones commented on staging/src/k8s.io/apiserver/pkg/cel/ip.go:
```diff
@@ -0,0 +1,92 @@
+	switch typeVal {
+	case ipAddrTypeValue:
+		return d
+	case types.TypeType:
```
Do you possibly want to consider conversion to string? I believe this will help with the `string.format` call

TristonianJones replied:
Rather that produce the string type name, can you convert back to a string representation?

JoelSpeed replied:
Ahh right I see, I looked at the upstream code for inspiration on this and ended up at the `type` type, I should have looked at a few more types to see what they did, can fix

TristonianJones commented on staging/src/k8s.io/apiserver/pkg/cel/library/cidr.go:
```diff
@@ -0,0 +1,274 @@
+// cidr('192.168.0.1/24') == cidr('192.168.0.1/24').masked() // returns false, CIDR was not in canonical format
+// cidr('192.168.0.0/16').preefixLength() // returns 16
+// cidr('::1/128').prefixLength() // returns 128
+func CIDR() cel.EnvOption {
```
As a high-level comment, is there anything in here which you wouldn't want to exposed as a canonical CEL extension? @JoelSpeed @jpbetz

TristonianJones commented on staging/src/k8s.io/apiserver/pkg/cel/library/cidr.go:
```diff
@@ -0,0 +1,274 @@
+}
+
+func (*cidrs) CompileOptions() []cel.EnvOption {
+	options := []cel.EnvOption{}
```
Don't forget to add `cel.Types(apiservercel.IPType, apiservercel.CIDRType)` as an environment option

JoelSpeed replied:
I don't actually know why we need that here, can you expand on why those are needed?

TristonianJones replied:
These are type identifiers which are needed if the type is referenced directly or as a field in a struct. If you don't provide the identifier expressions like the following should fail: `type(<ip>) == net.IP`, and references to ip types within fields won't resolve correctly.

JoelSpeed replied:
I tried to add a test `type(ip('192.168.0.1')) == net.IP` and I'm getting `"undeclared reference to 'net' (in container '')"`. I can see that the net.IP is registered in `revTypeMap` but when I dive deeper into the code, it's not registered as an ident, which is where it's being looked up. Any thoughts?

TristonianJones replied:
@JoelSpeed I think there might be a slight bug in the way top-level types are resolved during type-check. As a workaround, include the following line after you add the type: 

```
cel.Variable(apiservercel.IPType.TypeName(), cel.NewTypeTypeWithParam(apiservercel.IPType))
```

I've tested the fix within CEL and even if you leave the workaround in, it will just become a noop once the fix is made. I'll follow up with a CEL PR reference.

TristonianJones commented on staging/src/k8s.io/apiserver/pkg/cel/cidr.go:
```diff
@@ -0,0 +1,93 @@
+
+var (
+	CIDRObject    = decls.NewObjectType("kubernetes.CIDR")
+	cidrTypeValue = types.NewTypeValue("kubernetes.CIDR")
```
nit: use CIDRType everywhere instead as it is also a type instance

JoelSpeed commented:
I'm working on addressing the outstanding feedback today, going to go down the route [of this thread](https://github.com/kubernetes/kubernetes/pull/121912#discussion_r1410072741) and move back to `containsIP` and `containsCIDR` and also make a freestanding `ip.isCanonical` function that will make more sense than the `ip(<string>).wasCanonical()` that we currently have

TristonianJones commented on staging/src/k8s.io/apiserver/pkg/cel/library/cidr.go:
```diff
@@ -0,0 +1,276 @@
+
+func stringToCIDR(arg ref.Val) ref.Val {
+	s, ok := arg.Value().(string)
+	if !ok {
```
The CEL dispatching should guarantee the argument type. If you issue a call to this function where you wrap the argument in a dyn call like so `cidr(dyn(<arg>))`, this line should never be reached. I would add a couple tests if you want to make sure, but this should simplify these function implementations

TristonianJones replied:
@JoelSpeed this was a shift in how dispatching worked within CEL to improve ease of use. Early on, the CEL library put the burden of type-checking on the function implementer because it would speed up dispatch for a couple of exceptionally common call paths. We realized this was error prone, and then shifted to always validate the argument types prior to invocation to make it less error-prone. Early extensions in K8s use the old style, but it would be great to simplify here since the checks are redundant now, and tests will be a good way to ensure this is the case.

FYI: @cici37, @jpbetz

jpbetz replied:
> the checks are redundant

@TristonianJones this is good to know. I'm OK with leaving the check in for this PR (if only for consistency) and then simplifying as a follow up.

I've opened https://github.com/kubernetes/kubernetes/issues/122239 to track.

JoelSpeed committed:
Add tests for IP type

JoelSpeed commented:
Moving back to WIP, when I started adding tests, there are a couple that Tristan suggested that I can't get working, will need to dig into why but I can't work the undeclared test issue out at the moment. Final commit has WIP tests.

JoelSpeed committed:
Add tests for CIDR type

JoelSpeed commented:
Thanks for all the reviews so far folks, I think I've got this to a point where I'm happy with it. It covers the use cases that I (and I believe others) wanted and I've got the tests written up now.

PTAL again when you have a moment

jpbetz reviewed (approved):
/approve
/lgtm

Thanks @JoelSpeed!

jpbetz commented:
/retest

jpbetz commented:
/priority important-longterm
/triage accepted

jpbetz commented:
/release-note-edit ```release-note Add CEL library for IP Addresses and CIDRs. This will not be available for use until 1.31.```

jpbetz commented:
This just need a release note and it should merge.
